### ISSUE TITLE
API to access the shell type for a given terminal

### ISSUE THREAD

**Issue by karthiknadig** [microsoft/vscode#230165]
Currently there is no API available on the terminal that allows extensions to detect the shell type. This is needed for Python where there are shell specific activation scripts.

**Comment by gjsjohnmurray** [comment]
Does https://github.com/microsoft/vscode/issues/127374#issuecomment-870097462 point to a way?

**Comment by karthiknadig** [comment]
@gjsjohnmurray That is not available to Terminals created using `ExtensionTerminalOptions`.

**Comment by gjsjohnmurray** [comment]
There isn't a universal concept of "shell type" for a terminal contributed by an extension. However it's possible to define your own interface that extends `ExtensionTerminalOptions` so it can hold other properties that get set when the extension terminal gets created. These extra properties will be present in `Terminal.creationOptions` when a terminal is fetched from `window.activeTerminal` or `window.terminals[]`

**Comment by karthiknadig** [comment]
@gjsjohnmurray There is a `terminalShellType` context that is set for terminals, it should likely be set as a property on the terminal itself. https://github.com/microsoft/vscode/blob/b2f14985386e88f10275c4ee2203bb0f46346d14/src/vs/platform/terminal/common/terminal.ts#L149 

https://github.com/microsoft/vscode/blob/b2f14985386e88f10275c4ee2203bb0f46346d14/src/vs/workbench/contrib/terminal/common/terminalContextKey.ts#L85

The issue is that, when any extension or VS Code creates a terminal (DAP run-in-terminal for example) or the user click on the ➕ button to create terminals there is no consistent way to detect what shell it is using by examining the terminal instance. I understand that the shell type is a transient thing in a terminal. But, to provide better experience we need to know what it is at that moment for the terminal.

For Python extension, we need this for terminals created from any source, not just from python extension. The reason we need it is so we can configure the terminal to use the right python and packages from the correct locations. Typically, this is done by running a script provided by the python environment. These scripts are shell specific, and without this detail it leads to a bad terminal experience if we can't configure the environment correctly.

Currently we detect shells is by using the `Terminal.name` on terminal creation. This is limiting because, when some other source like DAP creates it that name is not set to the shell type, and that breaks detection. In cases where user clicks the ➕ the name is set to shell type.

**Comment by Tyriar** [comment]
@gjsjohnmurray we have some detection internally based on various factors like process name, executables in the process tree, etc. Exposing that would work for terminals created via extensions too, as long as they are backed by an executable, not a `Pseudoterminal` API object.

**Comment by Tyriar** [comment]
Proposal:

```ts
declare module 'vscode' {

	// https://github.com/microsoft/vscode/issues/230165

	/**
	 * Known terminal shell types.
	 */
	export enum TerminalShellType {
		Sh = 1,
		CommandPrompt,
		PowerShell,
		GitBash,
		Bash,
		Zsh,
		Python,
		Julia,
		NuShell,
		Fish,
		Csh,
		Ksh
	}

	// NOTE: State since this the shellType can change multiple times and this comes with an event.
	export interface TerminalState {
		/**
		 * The current detected shell type of the terminal. New shell types may be added in the
		 * future in which case they will be returned as a number that is not part of
		 * {@link TerminalShellType}.
		 * TODO: number is to prevent the breaking change when new enum members are added?
		 */
		readonly shellType: TerminalShellType | number | undefined;
	}

}
```

@karthiknadig any feedback?

**Comment by Tyriar** [comment]
The `TerminalShellType | number` approach seems ok for now as a means to prevent breaking in an enum that's likely changing in the future. Will probably get discussed more during finalization.

**Comment by Tyriar** [comment]
We should consider setting shellType based on the initial shell executable path so it's non-undefined initially.

**Comment by Tyriar** [comment]
Added @anthonykim1 since he's done a bunch with shell types at this point.

**Comment by Tyriar** [comment]
@anthonykim1 we need to keep this open as the API is not finalized

**Comment by Tyriar** [comment]
To discuss in API sync: Since supporting all shells is going to be pretty ugly, should we just use `string` as the type here and outline how that translates to the shell type in the jsdoc? For example it could say it translates to the standard binary of the shell type being run, like `'powershell'`, `'bash'`, etc. We don't even need to list them all out if we do this and we could also maybe pass through "shell types" that we aren't aware of if we're just passing the binary?

**Comment by Tyriar** [comment]
cc @meganrogge 👆

**Comment by Tyriar** [comment]
Also curious on your feedback @karthiknadig

**Comment by meganrogge** [comment]
Yes, I think that makes sense. Plus, it provides flexibility in case we want to add support for other shells in the future.

**Comment by karthiknadig** [comment]
+1 From initial usage in environments extension, it is harder to understand this when support is needed for shell types not already in this list. I prefer the `string` type for this API.

**Comment by Tyriar** [comment]
I was thinking we could also maybe just give the user the shell binary but we don't actually necessarily have that information currently. For Windows we walk the process tree and try to intelligently select the most inner shell so we can detect sub shells, for Mac and Linux simply look at the "terminal title" (via native code in node-pty):

https://github.com/microsoft/vscode/blob/1a75418ec72049f7583537755c8a536615ec782f/src/vs/platform/terminal/node/terminalProcess.ts#L124

So that approach wouldn't work without more effort (to do process list walking on Mac and Linux). Just standardizing on shell binary with a list of possibilities is the easiest way given what we have currently (eg. bash, zsh, pwsh, ...).

**Comment by jrieken** [comment]
* Terminal.state.shellType vs Terminal.shell.type (and Terminal.shell.integration)
* Type vs kind (naming)
* number vs string enum

**Comment by anthonykim1** [comment]
This API has already went through testing in: https://github.com/microsoft/vscode/issues/238888 

After some suggestions in API sync, now has been finalized: https://github.com/microsoft/vscode/pull/243274 

Main differences are the naming change from `shellType` to `shell` and the fact that the type is not enum but rather string. 
So instead of seeing number equivalent of the shell type, you would see the actual shell type as string name.


Just need quick verification to make sure string value of shell shows up:
Steps are listed in #238888 but will list them here:

1. Feel free to test on extension on your own.
2. Subscribing to window.onDidChangeTerminalState and watch the `shell` field.
3. Make sure you get the "correct" shell enum depending on which shell you open inside VS Code
4. feel free to test inner shell scenario too (such as opening `python` from `zsh`)

More description on vscode.d.ts itself:
https://github.com/microsoft/vscode/blob/89eac81178b9d9a221a1f9115ac3e9009e567559/src/vscode-dts/vscode.d.ts#L7750

Thank you.
